### PR TITLE
Verify if lvalue is in known arguments

### PR DESCRIPTION
**Pull request by yamatoRT**
Hello, 
This PR solve this issue : [#3047](https://github.com/robotframework/robotframework/issues/3047)
It's my first PR, don't hesitate to give me advices.

Regards,

### CODE REVIEW

**Comment by pekkaklarck** [comment]
Big thanks for the PR and sorry for review taking so long. I'm now concentrating on RF 3.1.2 and try to get it release at the end of April.

The fix itself looks very good. The only comment I have about the code is that the added empty line should be removed.

Tests also look good. We do, however, already have tests for named-only-arguments and the new test should be added there. The functionality ought to be pretty well tested so new tests are only needed for this particular bug. There are existing tests both for [user keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/user_keyword.robot) and for [library keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/python.robot) and I guess tests could be added to both.

**Comment by yamatoRT** [comment]
Hello,
thanks for you response. I moved my tests in [user keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/user_keyword.robot) and [library keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/python.robot). My modification create regressions, because i had this error after i was running test : 
```
 Unexpected keyword argumemt:
Expected: Keyword 'Kw Only Arg' got unexpected named argument 'invalid'.
Actual: Keyword 'Kw Only Arg' got positional argument after named arguments.
```
Then, i prefer consider all arguments in format 'name=value' after the first named argument as named arguments and not as positional arguments.

Moreover, i removed two typos in name of test case (argumemt become argument).

I'm waiting for your feedback.

Regards,

**Comment by pekkaklarck** [comment]
Two comments about the PR:

1. Most of the added tests are duplicate to existing tests. Only tests related to the actual change are needed. If there are missing tests, they can be added as a separate PR. Small changes like typo fixes are fine.

2. You are right that the original simple fix made error messages worse with cases like `nonex=value`. The added check ought to fix that . It also changes error messages in some other cases and related tests would need to be updated.

I've started to think is fixing this a good idea in general. It is somewhat annoying that you cannot nowadays should argument `foo=bar` be evaluated as a positional argument with literal value `foo=bar` or does it mean name argument `foo` with value `bar`. There have been some discussions about changing the syntax so that `foo=bar` would *always* be considered named argument syntax and literal value would require escaping like `foo\=bar`. That would be a big backwards incompatible change requiring long deprecation period and it may not be worth it. If we ever consider doing that, now allowing `foo=bar` as a literal value with named-only-arguments doesn't make much sense.

Let me think about this a bit more before continuing with this PR.

**Comment by pekkaklarck** [comment]
OK, let's do this. At the moment named-only arguments behave inconsistently compared to other arguments and that's not a good thing.

The current implementation is good, but I got some style related comments about the code. I'd also like tests to be enhanced. That basically means removing most of the added tests as they are duplicate to tests that already exist. I'll add more detailed review comments about these.

File: atest/testdata/keywords/named_only_args/KwOnlyArgs.py [hunk 1]
```diff
@@ -31,3 +31,7 @@ def all_arg_types(pos_req, pos_def='pd', *varargs,
     varargs = list(varargs)
     kwargs = ['%s=%s' % item for item in sorted(kwargs.items())]
     return '-'.join([pos_req, pos_def] + varargs + [kwo_req, kwo_def] + kwargs)
+
+
+def kw_two_args_with_varargs(*varargs, kwo='default', another='another'):
+    return '-'.join(varargs + (kwo, another,))
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/KwOnlyArgs.py]: I removed keywork. Now, I use  `Kw Only Arg ` and `Kw Only Arg With Varargs`.

File: atest/testdata/keywords/named_only_args/python.robot [hunk 2]
```diff
@@ -44,14 +44,22 @@ Missing multiple values
...
     Kw Only Arg    kwo=value    invalid=ooops    ooops=invalid
 
+Multiple unexpected keyword argument with inequality
+    [Documentation]    FAIL Keyword 'KwOnlyArgs.Kw Only Arg' got unexpected named arguments '<' and 'ooops'.
+    Kw Only Arg    kwo=value    <=ooops    ooops=invalid
+
+Multiple unexpected keyword argument with escaped inequality
+    [Documentation]    FAIL Keyword 'KwOnlyArgs.Kw Only Arg' got positional argument after named arguments.
+    Kw Only Arg    kwo=value    <\=ooops    ooops=invalid
+
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/python.robot]: I removed test case.

File: atest/testdata/keywords/named_only_args/python.robot [hunk 3]
```diff
@@ -75,3 +83,19 @@ With other arguments
     Should Be Equal    ${result}    p1-p2-p3-p4-k1-k2-k3=3-k4=4
     ${result} =    All Arg Types    k4=!!!    kwo_def=k2    k3=!    pos_req=p1    pos_def=p2    kwo_req=k1
     Should Be Equal    ${result}    p1-p2-k1-k2-k3=!-k4=!!!
+
+With only one named argument
+    Kw Two Args With Varargs    First arg    Second arg    Maybe a third    kwo=Actually, correct!
+
+With only one named argument but we have a escaped egal
+    Kw Two Args With Varargs    Let's try this again    1    <\=    2    kwo=Correct again!
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/python.robot]: I removed test case.

File: atest/testdata/keywords/named_only_args/python.robot [hunk 4]
```diff
@@ -75,3 +83,19 @@ With other arguments
...
+
+With only one named argument but we have a no escaped egal
+    Kw Two Args With Varargs    One more time    1    <=    2    kwo=This part fails
+
+With only one named argument but we have two no escaped egals
+    Kw Two Args With Varargs    One more time    a=1    <=    2    kwo=This part fails
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/python.robot]: I renamed test case. I changed value of kwo argument. I don't understand your last comment.

File: atest/testdata/keywords/named_only_args/user_keyword.robot [hunk 5]
```diff
@@ -80,6 +88,23 @@ With other arguments
...
+With only one named argument but we have a no escaped egal
+    Kw Two Args With Varargs    One more time    1    <=    2    kwo=This part fails
+
+With only one named argument but we have two no escaped egals
+    Kw Two Args With Varargs    One more time    a=1    <=    2    kwo=This part fails
+
+With two named arguments but with a error
+    [Documentation]  FAIL    Keyword 'Kw Two Args With Varargs' got positional argument after named arguments.
+    Kw Two Args With Varargs    One more time    another=1    2      <=     kwo=This part fails
+
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/user_keyword.robot]: I removed test case.

File: atest/testdata/keywords/named_only_args/user_keyword.robot [hunk 6]
```diff
@@ -115,3 +140,8 @@ All Arg Types
     ...    ${pos_req}    ${pos_def}    @{varargs}
     ...    ${kwo_req}    ${kwo_def}    @{kwargs}
     [Return]    ${result}
+
+Kw Two Args With Varargs
+    [Arguments]    @{varargs}    ${kwo}=default   ${another}=another
+    ${result} =    Catenate    SEPARATOR=-    @{varargs}    ${kwo}    ${another}
+    [Return]    ${result}
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/user_keyword.robot]: I removed keyword.

File: src/robot/running/arguments/argumentresolver.py [hunk 7]
```diff
@@ -63,13 +63,16 @@ def _is_dict_var(self, arg):
         return (is_string(arg) and arg[:2] == '&{' and arg[-1] == '}' and
                 VariableSplitter(arg).is_dict_variable())
 
-    def _is_named(self, arg, variables=None):
+    def _is_named(self, arg, have_even_named_argument, variables=None):
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply src/robot/running/arguments/argumentresolver.py]: I named argument `previous_named`.

File: src/robot/running/arguments/argumentresolver.py [hunk 8]
```diff
@@ -63,13 +63,16 @@ def _is_dict_var(self, arg):
...
         if not (is_string(arg) and '=' in arg):
             return False
         name, value = split_from_equals(arg)
+
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply src/robot/running/arguments/argumentresolver.py]: I removed empty line.

File: src/robot/running/arguments/argumentresolver.py [hunk 9]
```diff
@@ -63,13 +63,16 @@ def _is_dict_var(self, arg):
...
         if value is None:
             return False
-        if self._argspec.kwargs or self._argspec.kwonlyargs:
+        if self._argspec.kwargs or name in self._argspec.kwonlyargs:
+            return True
+        if have_even_named_argument and value is not None:
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply src/robot/running/arguments/argumentresolver.py]: Yes, my bad. I removed it.

File: atest/testdata/keywords/named_only_args/python.robot [hunk 10]
```diff
@@ -75,3 +83,19 @@ With other arguments
...
+
+With two named arguments but with a error
+    [Documentation]  FAIL    Keyword 'KwOnlyArgs.Kw Two Args With Varargs' got positional argument after named arguments.
+    Kw Two Args With Varargs    One more time    another=1    2      <=     kwo=This part fails
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/python.robot]: Yes, test already exists. I removed it.

File: atest/testdata/keywords/named_only_args/user_keyword.robot [hunk 11]
```diff
@@ -48,14 +48,22 @@ Missing multiple values
...
     Kw Only Arg    kwo=value    invalid=ooops    ooops=invalid
 
+Multiple unexpected keyword argument with inequality
+    [Documentation]    FAIL Keyword 'Kw Only Arg' got unexpected named arguments '<' and 'ooops'.
+    Kw Only Arg    kwo=value    <=ooops    ooops=invalid
+
+Multiple unexpected keyword argument with escaped inequality
+    [Documentation]    FAIL Keyword 'Kw Only Arg' got positional argument after named arguments.
+    Kw Only Arg    kwo=value    <\=ooops    ooops=invalid
+
```
[the remark this answers is not in the document]
Comment (yamatoRT) [reply atest/testdata/keywords/named_only_args/user_keyword.robot]: I removed test case.

**Comment by pekkaklarck** [comment]
Thanks for the changes! There are small things that could still be enhanced in tests, but I'll merge this now and do those changes myself afterwards.

**Comment by pekkaklarck** [comment]
Found a small bug in the implementation. It is possible to use `${var}=value` and in that case `${var}` should be resolved before checking is `name in argspec.kwonlyargs`. Resolving is already done before checking is `name in argspec.positional` and fixing the bug only requires doing that a bit earlier. I'll add tests and fix this myself.

**Pull request merged**